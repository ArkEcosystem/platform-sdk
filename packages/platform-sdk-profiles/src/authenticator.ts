import { Bcrypt } from "@arkecosystem/platform-sdk-crypto";

import { ProfileSetting } from "./profile.models";

export class Authenticator {
	readonly #profile;

	public constructor(profile) {
		this.#profile = profile;
	}

	public setPassword(password: string): void {
		this.#profile.settings().set(ProfileSetting.Password, Bcrypt.hash(password));
	}

	public verifyPassword(password: string): boolean {
		return Bcrypt.verify(this.#profile.settings().get(ProfileSetting.Password), password);
	}

	public changePassword(oldPassword: string, newPassword: string): void {
		const currentPassword: string | undefined = this.#profile.settings().get(ProfileSetting.Password);

		if (!currentPassword) {
			throw new Error("No password is set. Call [setPassword] instead.");
		}

		if (!Bcrypt.verify(currentPassword, oldPassword)) {
			throw new Error("The current password does not match.");
		}

		this.setPassword(newPassword);
	}
}
