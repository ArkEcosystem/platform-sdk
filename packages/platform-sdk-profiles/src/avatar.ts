import { createHash } from "crypto";
import MersenneTwister from "mersenne-twister";

export class Avatar {
	readonly #seed: string;
	readonly #colors: string[] = [
		"#742A2A",
		"#9B2C2C",
		"#E53E3E",
		"#F56565",
		"#FC8181",
		"#FEB2B2",
		"#C05621",
		"#DD6B20",
		"#ED8936",
		"#F6AD55",
		"#FBD38D",
		"#2C7A7B",
		"#319795",
		"#38B2AC",
		"#4FD1C5",
		"#4C51BF",
		"#5A67D8",
		"#667EEA",
		"#7F9CF5",
	];

	private constructor(seed: string) {
		this.#seed = seed;
	}

	public static make(seed: string): string {
		return new Avatar(seed).generate();
	}

	public generate(): string {
		const gradients: string[] = [];
		const shapesLength = 5;

		const seed = this.valueToSeed();
		const generator = new MersenneTwister(seed);

		for (let index = 0; index < shapesLength; index++) {
			const { angle, x, y } = this.calculateRandomPoints(generator);
			const color = this.getRandomColor(generator);
			gradients.push(
				`conic-gradient(
						from ${angle * 100}deg at ${x * 100}% ${y * 100}%,
						${color},
						calc(${index} * 100% / ${shapesLength}),
						transparent 0
					)`,
			);
		}

		return gradients.join(",");
	}

	private getRandomColor(generator: any): string {
		return this.#colors[Math.floor(this.#colors.length * generator.random())];
	}

	private calculateRandomPoints(generator: any) {
		const random = generator.random();
		const random2 = generator.random();

		const angle = Math.PI * 2 * random;
		const x = Math.abs(Math.sqrt(random2) * Math.cos(angle));
		const y = Math.abs(Math.sqrt(random2) * Math.sin(angle));

		return { angle, x, y };
	}

	private valueToSeed(): number {
		const hash = createHash("sha1")
			.update(Buffer.from(this.#seed, "utf-8"))
			.digest("latin1");

		// Change hash into a 32 bit int for our seed
		return (hash.charCodeAt(0) << 24) | (hash.charCodeAt(1) << 16) | (hash.charCodeAt(2) << 8) | hash.charCodeAt(3);
	}
}
