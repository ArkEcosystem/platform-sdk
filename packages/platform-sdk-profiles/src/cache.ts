import { DateTime } from "@arkecosystem/platform-sdk-intl";
import { SHA1 } from "bcrypto";
import dot from "dot-prop";

type CacheStore = Record<string, { expires_at: DateTime; value: unknown }>;

export class Cache {
	readonly #prefix: string;
	#cache: CacheStore = {};

	public constructor(prefix: string) {
		this.#prefix = prefix;
	}

	public all(): CacheStore {
		return this.#cache;
	}

	public keys(): string[] {
		return Object.keys(this.#cache);
	}

	public get<T>(key: string): T {
		const value: T | undefined = dot.get(this.#cache, this.getCacheKey(key));

		if (value === undefined) {
			throw new Error(`The [${key}] is an unknown cache value.`);
		}

		return value;
	}

	public set(key: string, value: unknown, ttl: number): void {
		dot.set(this.#cache, this.getCacheKey(key), {
			expires_at: DateTime.make().addSeconds(ttl),
			value,
		});

		this.checkExpiration(ttl);
	}

	public has(key: string): boolean {
		return dot.has(this.#cache, this.getCacheKey(key));
	}

	public forget(key: string): void {
		dot.delete(this.#cache, this.getCacheKey(key));
	}

	public flush(key: string): void {
		this.#cache = {};
	}

	private getCacheKey(value: unknown): string {
		return SHA1.digest(Buffer.from(`${this.#prefix}.${JSON.stringify(value)}`, "utf-8")).toString("hex");
	}

	private checkExpiration = (ttl: number) => {
		for (const [key, value] of Object.entries(this.all())) {
			if (value.expires_at.isBefore(DateTime.make())) {
				this.forget(key);
			}
		}

		setTimeout(this.checkExpiration, ttl * 1000, true);
	};
}
