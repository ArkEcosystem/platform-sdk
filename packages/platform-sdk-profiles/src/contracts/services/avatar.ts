export interface IAvatar {
    // make(seed: string): string;
}
