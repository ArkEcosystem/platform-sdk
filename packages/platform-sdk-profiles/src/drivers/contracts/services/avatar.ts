import { picasso } from "@vechain/picasso";
import mem from "mem";

export class Avatar {
	public static make(seed: string): string {
		return mem(picasso)(seed);
	}
}
