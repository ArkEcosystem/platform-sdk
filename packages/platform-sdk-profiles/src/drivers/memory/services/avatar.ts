import { picasso } from "@vechain/picasso";
import mem from "mem";
import { IAvatar } from "../../../contracts";

export class Avatar implements IAvatar {
	public static make(seed: string): string {
		return mem(picasso)(seed);
	}
}
