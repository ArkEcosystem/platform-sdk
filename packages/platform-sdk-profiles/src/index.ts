export * as Enums from "./enums";
export * from "./contact-repository";
export * from "./contracts";
export * from "./data";
export * from "./env";
export * from "./migrator";
export * from "./profile";
export * from "./profile-repository";
export * from "./settings";
export * from "./wallet-repository";
export * from "./wallet";
