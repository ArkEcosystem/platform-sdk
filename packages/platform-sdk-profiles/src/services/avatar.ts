import { picasso } from "@vechain/picasso";

export class Avatar {
	public static make(seed: string): string {
		return picasso(seed);
	}
}
