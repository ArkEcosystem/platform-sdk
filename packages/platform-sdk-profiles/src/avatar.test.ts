import "jest-extended";

import { Avatar } from "../src/avatar";

test("#profiles", async () => {
	expect(Avatar.make("Hello World")).toMatchSnapshot();
});
