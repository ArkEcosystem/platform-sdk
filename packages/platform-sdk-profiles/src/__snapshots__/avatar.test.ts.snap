// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#profiles 1`] = `"conic-gradient(from 277.8849158770073deg at 56.58010652464615% 21.47431676868284%, #7F9CF5, calc(0 * 100% / 5), transparent 0),conic-gradient(from 435.49648795276505deg at 14.088914554134977% 37.724875040067694%, #7F9CF5, calc(1 * 100% / 5), transparent 0),conic-gradient(from 46.37532015251623deg at 86.65369554846725% 43.33828585623631%, #DD6B20, calc(2 * 100% / 5), transparent 0),conic-gradient(from 119.50966823338247deg at 24.314758067076752% 61.644548474293416%, #7F9CF5, calc(3 * 100% / 5), transparent 0),conic-gradient(from 348.0437378849834deg at 48.0004499391454% 16.91716214791068%, #9B2C2C, calc(4 * 100% / 5), transparent 0)"`;
